flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/dynamic/DynamicRecordProcessor.java

```diff
@@ -0,0 +1,187 @@
+    boolean exists = tableCache.exists(data.tableIdentifier()).f0;
+    String foundBranch = exists ? tableCache.branch(data.tableIdentifier(), data.branch()) : null;
+
+    Tuple2<Schema, CompareSchemasVisitor.Result> foundSchema =
+        exists
+            ? tableCache.schema(data.tableIdentifier(), data.schema())
+            : TableDataCache.NOT_FOUND;
+
+    PartitionSpec foundSpec = exists ? tableCache.spec(data.tableIdentifier(), data.spec()) : null;
+
+    if (!exists
+        || foundBranch == null
+        || foundSpec == null
+        || foundSchema.f1 == CompareSchemasVisitor.Result.INCOMPATIBLE) {
+      if (immediateUpdate) {
+        Tuple3<Schema, CompareSchemasVisitor.Result, PartitionSpec> newData =
+            updater.update(data.tableIdentifier(), data.branch(), data.schema(), data.spec());
+        emit(collector, data, newData.f0, newData.f1, newData.f2);
+      } else {
+        int writerKey =
+            selector.getKey(
+                new DynamicKeySelector.Input(
+                    data,
+                    foundSchema.f0 != null ? foundSchema.f0 : data.schema(),
+                    foundSpec != null ? foundSpec : data.spec(),
+                    data.rowData()));
+        context.output(
+            updateStream,
+            new DynamicRecordInternal(
+                data.tableIdentifier().toString(),
+                data.branch(),
+                data.schema(),
+                data.spec(),
+                writerKey,
+                data.rowData(),
+                data.upsertMode(),
+                getEqualityFieldIds(data.equalityFields(), data.schema())));
+      }
+    } else {
+      emit(collector, data, foundSchema.f0, foundSchema.f1, foundSpec);
+    }
+  }
+
+  private void emit(
+      Collector<DynamicRecordInternal> out,
+      DynamicRecord data,
+      Schema schema,
+      CompareSchemasVisitor.Result result,
+      PartitionSpec spec) {
+    RowData rowData =
+        result == CompareSchemasVisitor.Result.SAME
+            ? data.rowData()
+            : RowDataEvolver.convert(data.rowData(), data.schema(), schema);
+    int writerKey = selector.getKey(new DynamicKeySelector.Input(data, schema, spec, rowData));
+    String tableName = data.tableIdentifier().toString();
+    out.collect(
+        new DynamicRecordInternal(
+            tableName,
+            data.branch(),
+            schema,
+            spec,
+            writerKey,
+            rowData,
+            data.upsertMode(),
+            getEqualityFieldIds(data.equalityFields(), schema)));
+  }
+
+  static List<Integer> getEqualityFieldIds(List<String> equalityFields, Schema schema) {
+    if (equalityFields == null || equalityFields.isEmpty()) {
+      if (!schema.identifierFieldIds().isEmpty()) {
+        return Lists.newArrayList(schema.identifierFieldIds());
+      } else {
+        return Collections.emptyList();
+      }
+    }
+    List<Integer> equalityFieldIds = Lists.newArrayList(equalityFields.size());
```

Comment:
This creates a list with one element equals to `equalityFields.size()`. I think the intention was to do something like `new ArrayList()`.